### PR TITLE
game resolution settings

### PR DESCRIPTION
**Pull request by Ladvace**

### CODE REVIEW

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/modals/InstanceManager/Overview.js  1
- src/common/modals/Settings/components/Java.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5329977&bid=17444954)

File: src/app/desktop/utils/index.js [hunk 1]
```diff
@@ -291,6 +292,8 @@ export const getJVMArguments112 = (
   args.push(`-Djava.library.path="${path.join(instancePath, 'natives')}"`);
 
   args.push(mcJson.mainClass);
+  if (resolution)
+    args.push(` --width ${resolution.width} --height ${resolution.height}`);
```
Comment (blarfoon) [inline src/app/desktop/utils/index.js]: Why the empty space? Also, don't push 2 different parameters into one string

File: src/app/desktop/utils/index.js [hunk 2]
```diff
@@ -368,13 +372,16 @@ export const getJVMArguments113 = (
   //   args.push("-Xdock:name=instancename");
   //   args.push("-Xdock:icon=instanceicon");
   // }
+  console.log('RES', resolution);
```
Comment (blarfoon) [inline src/app/desktop/utils/index.js]: Remove

File: src/app/desktop/utils/index.js [hunk 3]
```diff
@@ -368,13 +372,16 @@ export const getJVMArguments113 = (
...
 
   args.push(`-Xmx${memory}m`);
   args.push(`-Xms${memory}m`);
   args.push(`-Dminecraft.applet.TargetDirectory="${instancePath}"`);
   args.push(...jvmOptions);
 
   args.push(mcJson.mainClass);
+  if (resolution)
+    args.push(` --width ${resolution.width} --height ${resolution.height}`);
```
Comment (blarfoon) [inline src/app/desktop/utils/index.js]: Same as above

File: src/common/modals/InstanceManager/Overview.js [hunk 4]
```diff
@@ -86,6 +102,11 @@ const Overview = ({ instanceName }) => {
     config?.javaArgs
   );
   const [newName, setNewName] = useState(instanceName);
+  const [height, setHeight] = useState(config?.resolution?.height);
+  const [width, setWidth] = useState(config?.resolution?.width);
+  const [gameResolutionSwitch, setGameResolutionSwitch] = useState(
+    config?.resolution?.height || config?.resolution?.width
+  );
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: Actually we want to make sure that both are set, since one can't exist without the other

File: src/common/modals/InstanceManager/Overview.js [hunk 5]
```diff
@@ -107,6 +128,15 @@ const Overview = ({ instanceName }) => {
...
 
+  const updateGameResolution = (h, w) => {
+    dispatch(
+      updateInstanceConfig(instanceName, prev => ({
+        ...prev,
+        resolution: { height: h, width: w }
+      }))
+    );
+  };
+
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js:139]: Does this also remove them in case the switch is switched off?

File: src/common/modals/InstanceManager/Overview.js [hunk 6]
```diff
@@ -138,6 +168,67 @@ const Overview = ({ instanceName }) => {
...
+                    setWidth(w);
+                    dispatch(
+                      updateInstanceConfig(instanceName, prev => ({
+                        ...prev,
+                        resolution: {
+                          height,
+                          w
+                        }
+                      }))
+                    );
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: Don't repeat the code

File: src/common/modals/InstanceManager/Overview.js [hunk 7]
```diff
@@ -138,6 +168,67 @@ const Overview = ({ instanceName }) => {
...
+                  onChange={e => {
+                    const h = e.target.value;
+                    setHeight(h);
+                    dispatch(
+                      updateInstanceConfig(instanceName, prev => ({
+                        ...prev,
+                        resolution: {
+                          h,
+                          width
+                        }
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: Don't repeat the code

File: src/common/modals/Settings/components/Java.js [hunk 8]
```diff
@@ -81,12 +104,20 @@ const marks = {
...
   const javaMemory = useSelector(state => state.settings.java.memory);
   const javaPath = useSelector(_getJavaPath);
   const customJavaPath = useSelector(state => state.settings.java.path);
+  const customResolution = useSelector(state => state.settings.java.resolution);
   const dispatch = useDispatch();
 
+  useEffect(() => {
+    if (customResolution.height) setHeight(customResolution.height);
+    if (customResolution.width) setWidth(customResolution.width);
+  }, []);
```
Comment (blarfoon) [inline src/common/modals/Settings/components/Java.js]: Don't use the useEffect, assign them directly on states declaration

File: src/common/reducers/actions.js [hunk 9]
```diff
@@ -1574,6 +1575,9 @@ export function launchInstance(instanceName) {
 
     const javaArguments = (javaArgs !== undefined ? javaArgs : args).split(' ');
     const javaMem = javaMemory !== undefined ? javaMemory : memory;
+    const resol = resolution !== undefined ? resolution : gameResolution;
+
+    console.log('REE', resol, resolution, javaMemory, gameResolution, memory);
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: Remove

File: src/common/reducers/actions.js [hunk 10]
```diff
@@ -1574,6 +1575,9 @@ export function launchInstance(instanceName) {
 
     const javaArguments = (javaArgs !== undefined ? javaArgs : args).split(' ');
     const javaMem = javaMemory !== undefined ? javaMemory : memory;
+    const resol = resolution !== undefined ? resolution : gameResolution;
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: What if it's null

File: src/common/reducers/actions.js [hunk 11]
```diff
@@ -1474,10 +1474,11 @@ export function launchInstance(instanceName) {
     const account = _getCurrentAccount(state);
     const librariesPath = _getLibrariesPath(state);
     const assetsPath = _getAssetsPath(state);
-    const { memory, args } = state.settings.java;
-    const { modloader, javaArgs, javaMemory } = _getInstance(state)(
+    const { memory, args, resolution: gameResolution } = state.settings.java;
+    const { modloader, javaArgs, javaMemory, resolution } = _getInstance(state)(
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: Use more specific and understandable names

File: src/common/reducers/settings/actionTypes.js [hunk 12]
```diff
@@ -5,6 +5,7 @@ export const UPDATE_SHOW_CHANGELOG = 'settings/UPDATE_SHOW_CHANGELOG';
 export const UPDATE_CONCURRENT_DOWNLOADS =
   'settings/UPDATE_CONCURRENT_DOWNLOADS';
 
+export const UPDATE_RESOLUTION = 'settings/UPDATE_RESOLUTION';
```
Comment (blarfoon) [inline src/common/reducers/settings/actionTypes.js]: rename into `UPDATE_MINECRAFT_RESOLUTION`

File: src/common/reducers/settings/reducers.js [hunk 13]
```diff
@@ -85,6 +85,11 @@ function java(
       return { ...state, memory: action.memory };
     case ActionTypes.UPDATE_JAVA_PATH:
       return { ...state, path: action.path };
+    case ActionTypes.UPDATE_RESOLUTION:
+      return {
+        ...state,
+        resolution: action.resolution
+      };
```
Comment (blarfoon) [inline src/common/reducers/settings/reducers.js]: This is not directly related to java. We might want to create a new tab in settings called `Minecraft Settings`

**Review by blarfoon — CHANGES_REQUESTED** [review]

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/modals/Settings/components/Java.js  2
- src/common/modals/InstanceManager/Overview.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5329977&bid=17444954)

File: src/common/reducers/settings/reducers.js [hunk 14]
```diff
@@ -70,6 +70,18 @@ function showNews(state = true, action) {
...
+    case ActionTypes.UPDATE_MINECRAFT_RESOLUTION:
+      return {
+        ...state,
+        resolution: action.resolution
+      };
+    default:
+      return state;
+  }
+}
+
```
Comment (blarfoon) [inline src/common/reducers/settings/reducers.js:84]: How do you remove it when the user removes it?

**Review by blarfoon — CHANGES_REQUESTED** [review]

File: src/common/modals/InstanceManager/Overview.js (hunk 5 above)
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js:139]: please call parameters width and height and use property value shorthand

File: src/common/modals/InstanceManager/Overview.js [hunk 15]
```diff
@@ -138,6 +168,70 @@ const Overview = ({ instanceName }) => {
...
+                    updateInstanceConfig(instanceName, prev =>
+                      omit(prev, ['resolution'])
+                    )
+                  );
+                } else if (v) {
+                  updateGameResolution(600, 800);
+                  setHeight(600);
+                  setWidth(800);
+                }
+              }}
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: It's obvious that it's v if it's not !v. Remove else if

File: src/common/modals/InstanceManager/Overview.js [hunk 16]
```diff
@@ -138,6 +168,70 @@ const Overview = ({ instanceName }) => {
...
+          {gameResolutionSwitch && (
+            <ResolutionInputContainer>
+              <div>
+                <Input
+                  placeholder="height"
+                  value={height}
+                  onChange={e => {
+                    const h = parseInt(e.target.value, 10);
+                    setHeight(h);
+                    setWidth(width);
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: What's the point of this line?

File: src/common/modals/InstanceManager/Overview.js [hunk 17]
```diff
@@ -138,6 +168,70 @@ const Overview = ({ instanceName }) => {
...
+                    );
+                  }}
+                />
+                &nbsp;X&nbsp;
+                <Input
+                  placeholder="width"
+                  value={width}
+                  onChange={e => {
+                    const w = parseInt(e.target.value, 10);
+                    setHeight(height);
```
Comment (blarfoon) [inline src/common/modals/InstanceManager/Overview.js]: What's the point of this line?

File: src/common/reducers/actions.js [hunk 18]
```diff
@@ -1583,6 +1591,7 @@ export function launchInstance(instanceName) {
       mcJson,
       account,
       javaMem,
+      resol,
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: Resol is not a decent variable name

File: src/common/reducers/actions.js [hunk 19]
```diff
@@ -1607,6 +1616,7 @@ export function launchInstance(instanceName) {
         mcJson,
         account,
         javaMem,
+        resol,
```
Comment (blarfoon) [inline src/common/reducers/actions.js]: Resol is not a decent variable name

File: src/common/reducers/settings/reducers.js (hunk 14 above)
Comment (blarfoon) [reply src/common/reducers/settings/reducers.js:84]: Up

File: src/common/reducers/settings/actions.js [hunk 20]
```diff
@@ -54,6 +54,15 @@ export function updateShowChangelog(show) {
...
 
+export function updateResolution(resolution) {
+  return dispatch => {
+    dispatch({
+      type: ActionTypes.UPDATE_MINECRAFT_RESOLUTION,
+      resolution
+    });
+  };
+}
+
```
Comment (blarfoon) [inline src/common/reducers/settings/actions.js:65]: updateMinecraftResolution

File: src/common/modals/Settings/components/Java.js [hunk 21]
```diff
@@ -173,6 +212,52 @@ export default function MyAccountPreferences() {
...
+              }}
+            />
+            &nbsp;X&nbsp;
+            <Input
+              placeholder="width"
+              value={width}
+              onChange={e => {
+                const w = parseInt(e.target.value, 10);
+                setWidth(w);
+                dispatch(updateResolution({ height, pane: w }));
```
Comment (blarfoon) [inline src/common/modals/Settings/components/Java.js]: what is pane?

File: src/common/modals/Settings/components/Java.js [hunk 22]
```diff
@@ -81,12 +104,28 @@ const marks = {
...
+  useEffect(() => {
+    if (customResolution.width && customResolution.height) {
+      setWidth(customResolution.width);
+      setHeight(customResolution.height);
+    } else {
+      setWidth(800);
+      setHeight(600);
+      dispatch(updateResolution({ height: 600, width: 800 }));
+    }
+  }, []);
```
Comment (blarfoon) [inline src/common/modals/Settings/components/Java.js]: This should not be here. Put this in state declaration

**Review by blarfoon — CHANGES_REQUESTED** [review]

**Comment by killpowa** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/modals/InstanceManager/Overview.js  2
- src/common/modals/Settings/components/Java.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5329977&bid=17444954)

File: src/common/modals/InstanceManager/Overview.js (hunk 5 above)
Comment (Ladvace) [reply src/common/modals/InstanceManager/Overview.js:139]: height and width are already declared in the upper scope, that's why I did this

**Comment by blarfoon** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/common/modals/Settings/components/Java.js  2
- src/common/modals/InstanceManager/Overview.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/gorilla-devs/GDLauncher/pullRequest?prid=5329977&bid=17444954)
